File: HueQuickStartApps/HueQuickStartApp-iOS/HueQuickStartApp/Bridge/Discovery/PHBridgeSelectionViewController.m
/*******************************************************************************
 Copyright (C) 2017 Philips Lighting Holding B.V.
 All Rights Reserved.
 ********************************************************************************/

#import "PHBridgeSelectionViewController.h"
#import "PHAppDelegate.h"

static NSString* PHBridgeSelectionCellIdentifier = @"DiscoveredBridgeCell";

@interface PHBridgeSelectionViewController ()

@property (nonatomic, strong) NSArray<PHSBridgeDiscoveryResult *>* discoveredBridges;

@property (nonatomic, weak) id<PHBridgeSelectionViewControllerDelegate> delegate;

@end

@implementation PHBridgeSelectionViewController

- (instancetype)initWithDiscoveredBridges:(NSDictionary<NSString *, PHSBridgeDiscoveryResult *> *)bridges andDelegate:(id<PHBridgeSelectionViewControllerDelegate>)delegate {
    self = [super init];

    if (self) {
        self.discoveredBridges = [self bridgesDictionaryToArray:bridges];
        
        self.delegate = delegate;
    }
    
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self setupNavigationBar];
    
    [self.tableView reloadData];
}

- (IBAction)refreshButtonClicked:(id)sender {
    [self.delegate refreshBridges];
}

- (NSArray *)bridgesDictionaryToArray:(NSDictionary<NSString *, PHSBridgeDiscoveryResult *> *)bridgesDictionary {
    // Add sorting here...
    
    return [bridgesDictionary allValues];
}

#pragma mark - UI

- (void)setupNavigationBar {
    self.navigationItem.hidesBackButton = YES;
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemRefresh
                                                                                           target:self
                                                                                           action:@selector(refreshButtonClicked:)];
    
    self.navigationItem.title = NSLocalizedString(@"Bridges", @"Title navigation found bridges screen");
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.discoveredBridges.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:PHBridgeSelectionCellIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle
                                      reuseIdentifier:PHBridgeSelectionCellIdentifier];
    }
    
    PHSBridgeDiscoveryResult* discoveredBridge = self.discoveredBridges[indexPath.row];
    
    cell.textLabel.text = discoveredBridge.ipAddress;
    cell.detailTextLabel.text = discoveredBridge.uniqueId;
    
    return cell;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    return NSLocalizedString(@"Please select a bridge to connect to", @"Header bridge selection view");
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    PHSBridgeDiscoveryResult* discoveredBridge = self.discoveredBridges[indexPath.row];
    
    [self.delegate bridgeSelectedWithIpAddress:discoveredBridge.ipAddress andUniqueId:discoveredBridge.uniqueId];
}

@end

File: HueQuickStartApps/HueQuickStartApp-iOS/HueQuickStartAppSwift/Bridge/Discovery/PHBridgeSelectionViewController.swift
/*******************************************************************************
 Copyright (C) 2017 Philips Lighting Holding B.V.
 All Rights Reserved.
 ********************************************************************************/

import Foundation
import UIKit

protocol PHBridgeSelectionViewControllerSegueCoordinator : SegueCoordinator {
    func bridgeSelectionController(_ controller:PHBridgeSelectionViewController, didSelectBridge bridge:PHBridgeInfo)
    func bridgeSelectionControllerDidRefreshBridges(_ controller:PHBridgeSelectionViewController)
}

let PHBridgeSelectionCellIdentifier:String = "DiscoveredBridgeCell";

extension UITableViewCell {
    func setBridgeInfo(_ bridgeInfo:PHBridgeInfo) {
        self.textLabel?.text = bridgeInfo.ipAddress
        self.detailTextLabel?.text = bridgeInfo.uniqueId
    }
}

class PHBridgeSelectionViewController : UITableViewController, NavigationHelping {
    var bridgeList:[PHBridgeInfo] = []
    
    lazy var segueCoordinator:PHBridgeSelectionViewControllerSegueCoordinator = {
       return HueQuickStartAppSegueCoordinator()
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        configureNavigation()
    }
    
    @IBAction func refreshButtonAction() {
        self.segueCoordinator.bridgeSelectionControllerDidRefreshBridges(self)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        self.segueCoordinator.coordinate(segue:segue, sender:sender)
    }
}

private typealias PHBridgeSelectionViewControllerTableDatasource = PHBridgeSelectionViewController
extension PHBridgeSelectionViewControllerTableDatasource {
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.bridgeList.count
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell:UITableViewCell = tableView.dequeueReusableCell(withIdentifier:PHBridgeSelectionCellIdentifier) ?? UITableViewCell(style: .subtitle, reuseIdentifier: PHBridgeSelectionCellIdentifier)
        
        cell.setBridgeInfo(bridgeList[indexPath.row])
        return cell
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        return NSLocalizedString("Please select a bridge to connect to", comment:"Header bridge selection view")
    }
}

private typealias PHBridgeSelectionViewControllerTableDelegate = PHBridgeSelectionViewController
extension PHBridgeSelectionViewControllerTableDelegate {
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        segueCoordinator.bridgeSelectionController(self, didSelectBridge:self.bridgeList[indexPath.row])
    }
}

private typealias PHBridgeSelectionViewControllerNavigable = PHBridgeSelectionViewController
extension PHBridgeSelectionViewControllerNavigable : NavigableController {
    var screenTitle: String {
        return NSLocalizedString("HueQuickStartApp", comment:"Title navigation light control screen")
    }
    
    var backButtonHidden: Bool {
        return true
    }
    
    var navigationBarHidden: Bool {
        return false
    }
    
    var rightNavigationButton: UIBarButtonItem {
        return UIBarButtonItem(barButtonSystemItem: .refresh, target: self, action: #selector(refreshButtonAction))
    }
}
